File: full_case_study_solns/_f_planar_vtol/python_old/hw12/VTOLParamHW12.py
# VTOL Parameter File
import numpy as np
import control as cnt
import sys
sys.path.append('..')  # add parent directory
import VTOLParam as P


####################################################
#                 State Space
####################################################
# tuning parameters
wn_h    = 1.0
zeta_h  = 0.707
wn_z    = 0.9905
zeta_z  = 0.707
wn_th   = 13.3803
zeta_th = 0.707
integrator_h = -1.0
integrator_z = -1.0

# State Space Equations
A_lon = np.matrix([[0.0, 1.0],
                   [0.0, 0.0]])
B_lon = np.matrix([[0.0],
                   [1.0/(P.mc+2.0*P.mr)]])
C_lon = np.matrix([[1.0, 0.0]])
A_lat = np.matrix([[0.0, 0.0, 1.0, 0.0],
                   [0.0, 0.0, 0.0, 1.0],
                   [0.0, -P.Fe/(P.mc+2.0*P.mr), -(P.mu/(P.mc+2.0*P.mr)), 0.0],
                   [0.0, 0.0, 0.0, 0.0]])
B_lat = np.matrix([[0.0],
                   [0.0],
                   [0.0],
                   [1.0/(P.Jc+2*P.mr*P.d**2)]])
C_lat = np.matrix([[1.0, 0.0, 0.0, 0.0],
                   [0.0, 1.0, 0.0, 0.0]])

# form augmented system
A1_lon = np.matrix([[0.0, 1.0, 0.0],
                    [0.0, 0.0, 0.0],
                    [-1.0, 0.0, 0.0]])
B1_lon = np.matrix([[0.0],
                    [1.0],
                    [0.0]])
A1_lat = np.matrix([[0.0, 0.0, 1.0, 0.0, 0.0],
                    [0.0, 0.0, 0.0, 1.0, 0.0],
                    [0.0, -P.Fe/(P.mc+2.0*P.mr), -(P.mu/(P.mc+2.0*P.mr)), 0.0, 0.0],
                    [0.0, 0.0, 0.0, 0.0, 0.0],
                    [-1.0, 0.0, 0.0, 0.0, 0.0]])
B1_lat = np.matrix([[0.0],
                    [0.0],
                    [0.0],
                    [1.0/(P.Jc+2*P.mr*P.d**2)],
                    [0.0]])

# gain calculation
des_char_poly_lon = np.convolve([1.0, 2.0*zeta_h*wn_h, wn_h**2],
                                np.poly(integrator_h))
des_poles_lon = np.roots(des_char_poly_lon)

des_char_poly_lat = np.convolve(
                        np.convolve([1.0, 2.0*zeta_z*wn_z, wn_z**2],
                                    [1.0, 2.0*zeta_th*wn_th, wn_th**2]),
                        np.poly(integrator_z))
des_poles_lat = np.roots(des_char_poly_lat)


# Compute the gains if the system is controllable
if np.linalg.matrix_rank(cnt.ctrb(A1_lon, B1_lon)) != 3:
    print("The longitudinal system is not controllable")
else:
    K1_lon = cnt.acker(A1_lon, B1_lon, des_poles_lon)
    K_lon = np.matrix([K1_lon.item(0), K1_lon.item(1)])
    ki_lon = K1_lon.item(2)

if np.linalg.matrix_rank(cnt.ctrb(A1_lat, B1_lat)) != 5:
    print("The lateral system is not controllable")
else:
    K1_lat = cnt.acker(A1_lat, B1_lat, des_poles_lat)
    K_lat = np.matrix([K1_lat.item(0), K1_lat.item(1), K1_lat.item(2), K1_lat.item(3)])
    ki_lat = K1_lat.item(4)

print('K_lon: ', K_lon)
print('ki_lon: ', ki_lon)
print('K_lat: ', K_lat)
print('ki_lat: ', ki_lat)

File: full_case_study_solns/_f_planar_vtol/matlab/hw12/VTOLParamHW12.m
% inverted ballbeam - parameter file for hw8
addpath ./.. % adds the parent directory to the path
VTOLParam % general ballbeam parameters

% turning parameters
wn_h    = 1;
zeta_h  = 0.707;
wn_z    = 0.9905;
zeta_z  = 0.707;
wn_th   = 13.3803;
zeta_th = 0.707;
integrator_h = -1;
integrator_z = -1;

% equilibrium force and constraints
P.Fe = (P.mc+2*P.mr)*P.g;
P.Ftildemax = 2*P.fmax - P.Fe;
P.taumax = (P.fmax-P.Fe/2)/P.d;

% state space design
A_lon = [...
    0, 1;...
    0, 0;...
    ];
B_lon = [0; 1/(P.mc+2*P.mr)];
C_lon = [1, 0];
A_lat = [...
    0, 0, 1, 0;...
    0, 0, 0, 1;...
    0, -(P.Fe/(P.mc+2*P.mr)), -(P.mu/(P.mc+2*P.mr)), 0;...
    0, 0, 0, 0;...
    ];
B_lat = [0;0;0;1/(P.Jc+2*P.mr*P.d^2)];
C_lat = [1, 0, 0, 0; 0, 1, 0, 0];

% form augmented system
A1_lon = [A_lon, zeros(2,1); -C_lon, 0];
B1_lon = [B_lon; 0];
Cout_lat = [1, 0, 0, 0];
A1_lat = [A_lat, zeros(4,1); -Cout_lat, 0];
B1_lat = [B_lat; 0];

des_char_poly_lon = conv([1,2*zeta_h*wn_h,wn_h^2],...
                          poly(integrator_h));
des_poles_lon = roots(des_char_poly_lon);
des_char_poly_lat = conv(conv([1,2*zeta_z*wn_z,wn_z^2],...
                              [1,2*zeta_th*wn_th,wn_th^2]),...
                         poly(integrator_z));
des_poles_lat = roots(des_char_poly_lat);

% gains for longitudinal system
% is the system controllable?
if rank(ctrb(A1_lon,B1_lon))~=3
    disp('System Not Controllable'); 
else % if so, compute gains
    K1_lon   = place(A1_lon,B1_lon,des_poles_lon); 
    P.K_lon  = K1_lon(1:2);
    P.ki_lon = K1_lon(3);
end

% gains for lateral system
% is the system controllable?
if rank(ctrb(A1_lat,B1_lat))~=5
    disp('System Not Controllable'); 
else % if so, compute gains
    K1_lat   = place(A1_lat,B1_lat,des_poles_lat); 
    P.K_lat  = K1_lat(1:4);
    P.ki_lat = K1_lat(5);
end

sprintf('K_lat: (%f, %f, %f, %f)\nki_lat: %f\nK_lon: (%f, %f)\nki_lon: %f\n',...
    P.K_lat(1), P.K_lat(2), P.K_lat(3), P.K_lat(4), P.ki_lat,...
    P.K_lon(1), P.K_lon(2), P.ki_lon)

